src/lib/Libifl/pbsD_connect.c

```diff
@@ -470,12 +349,186 @@ __pbs_connect_extend(char *server, char *extend_data)
 	 * Disable Nagle's algorithm on the TCP connection to server.
 	 * Nagle's algorithm is hurting cmd-server communication.
 	 */
-	if (pbs_connection_set_nodelay(sock) == -1) {
-		closesocket(sock);
+	if (pbs_connection_set_nodelay(sd) == -1) {
+		CLOSESOCKET(sd);
 		pbs_errno = PBSE_SYSTEM;
 		return -1;
 	}
 
+	return sd;
+}
+
+/**
+ * @brief	Get the array of connections to all servers
+ *
+ * @param	void
+ *
+ * @return	void
+ * @retval	!NULL - success
+ * @retval	NULL - error
+ *
+ * @par Side Effects:
+ *	None
+ *
+ * @par MT-safe: Yes
+ */
+void *
+get_conn_servers(void)
+{
+	svr_conn_t *conn_arr = NULL;
+
+	conn_arr = pthread_getspecific(psi_key);
+	if (conn_arr == NULL && pbs_conf.psi != NULL) {
+		int num_svrs;
+		int i;
+
+		num_svrs = get_num_servers();
+		conn_arr = calloc(num_svrs, sizeof(svr_conn_t));
+		if (conn_arr == NULL) {
+			pbs_errno = PBSE_SYSTEM;
+			return NULL;
+		}
+
+		for (i = 0; i < num_svrs; i++) {
+			strcpy(conn_arr[i].name, pbs_conf.psi[i].name);
+			conn_arr[i].port = pbs_conf.psi[i].port;
+			conn_arr[i].sd = -1;
+			conn_arr[i].secondary_sd = -1;
+			conn_arr[i].state = SVR_CONN_STATE_DOWN;
+		}
+
+		pthread_setspecific(psi_key, conn_arr);
+	}
+
+	return conn_arr;
+}
+
+
+/**
+ * @brief	Helper function for connect_to_servers to connect to a particular server
+ *
+ * @param[in]		idx - array index for the server to connect to
+ * @param[in,out]	conn_arr - array of svr_conn_t
+ * @param[in]		extend_data - any additional data relevant for connection
+ *
+ * @return	int
+ * @retval	-1 for error
+ * @retval	fd of connection
+ */
+static int
+connect_to_server(int idx, svr_conn_t *conn_arr, char *extend_data)
+{
+	if (conn_arr[idx].state != SVR_CONN_STATE_UP) {
+		if ((conn_arr[idx].sd = tcp_connect(conn_arr[idx].name, conn_arr[idx].port, extend_data)) != -1) {
+			conn_arr[idx].state = SVR_CONN_STATE_UP;
+			pbs_client_thread_lock_conntable();
+			add_connection(conn_arr[idx].sd);
```

Comment:
This seems wrong to me, this func was made only for conn_table.c (the one file who manipulates entries in connection table).
Also you don't need to do this, as tcp_connect calls to DIS routines (aka sends PBS_BATCH_Connect request and does authentication) which means given sd is already part of connection table. Here we are adding it 2nd time (although it won't add but it will just reset ch_errtxt and ch_errno of entry).

I would suggest, please revert add_connection() to static and remove this (along with lock/unlock of conntable).

Reply:
I remember that there were issues without this call, i can try removing it to see what happens.

Reply:
> Also you don't need to do this, as tcp_connect calls to DIS routines (aka sends PBS_BATCH_Connect request and does authentication) which means given sd is already part of connection table. Here we are adding it 2nd time (although it won't add but it will just reset ch_errtxt and ch_errno of entry).

Can you please tell me where it's added the first time ?

Reply:
It will be added at `__pbs_connect_extend()`->`encode_DIS_ReqHdr()`->`diswui()` (aka `diswul()`)->`dis_puts()`->`dis_get_writebuf()`->`transport_get_chan()` (aka `tcp_get_chan()` for tcp connection)->`get_conn_chan()`->`get_connection()`->`add_connection()`